### PR TITLE
Datagen API

### PR DESCRIPTION
**Pull request by robertbastian**
Part of #1188

### CODE REVIEW

File: provider/datagen/Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ include = [
     "LICENSE",
     "README.md",
 ]
+default-run = "icu4x-datagen"
```
Comment (Manishearth) [inline provider/datagen/Cargo.toml:25]: TIL!

File: provider/datagen/src/lib.rs [hunk 2]
```diff
@@ -2,288 +2,219 @@
...
+    Ok(keys(&BufReader::new(file).lines().collect::<std::io::Result<Vec<String>>>()?))
+}
 
-                $(<$constructor>::from(&__source)),+,
-            ]
-        )
-    }};
+/// The output format.
+#[non_exhaustive]
+pub enum Out {
```
Comment (sffc) [inline provider/datagen/src/lib.rs:121]: Praise: I like this abstraction

File: provider/datagen/src/lib.rs [hunk 3]
```diff
@@ -2,288 +2,219 @@
...
+///
+/// The argument are used as follows:
+/// * `locales`: If this is present, only locales that are either `und` or
+///   contained (strictly, i.e. `en` != `en-US`) in the slice will be generated.
+///   Otherwise, all locales supported by the source data will be generated.
+/// * `keys`: The keys for which to generate data. See [`get_all_keys()`].
+/// * `sources`: The underlying source data. CLDR and/or uprops data can be missing if no
+///   requested key requires them. Otherwise a error will be returned that can be identified
+///   with [`is_missing_cldr_error`] or [`is_missing_uprops_error`].
+/// * `out`: See the documentation on [`Out`]
```
Comment (Manishearth) [inline provider/datagen/src/lib.rs:151]: nit: should probably still say something like "details on the output format and location. see documentation on Out"

**Review by Manishearth — APPROVED** [review]

File: provider/datagen/src/lib.rs [hunk 4]
```diff
@@ -2,288 +2,219 @@
...
+    /// Output to a file system tree
+    Fs {
+        /// The root path.
+        output_path: PathBuf,
+        /// The serialization format. See [icu_provider_fs::export::serializers].
+        serializer: Box<dyn serializers::AbstractSerializer + Sync>,
+        /// Whether to overwrite existing data.
+        overwrite: bool,
+    },
+    /// Output as a postcard blob to the given sink.
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply provider/datagen/src/lib.rs:137]: Yeah we don't do JSON blobs

**Review by Manishearth — DISMISSED** [review]

File: provider/datagen/src/transform/cldr/serde/aliases.rs [hunk 5]
```diff
@@ -18,16 +19,16 @@ pub struct Replacement<T> {
...
-    #[serde(with = "tuple_vec_map", rename = "scriptAlias")]
-    pub script_aliases: Vec<(TinyAsciiStr<4>, Replacement<TinyAsciiStr<4>>)>,
-    #[serde(with = "tuple_vec_map", rename = "territoryAlias")]
-    pub region_aliases: Vec<(TinyAsciiStr<3>, Replacement<String>)>,
-    #[serde(with = "tuple_vec_map", rename = "variantAlias")]
-    pub variant_aliases: Vec<(TinyAsciiStr<8>, Replacement<TinyAsciiStr<8>>)>,
-    #[serde(with = "tuple_vec_map", rename = "subdivisionAlias")]
-    pub subdivision_aliases: Vec<(TinyAsciiStr<7>, Replacement<String>)>,
+    #[serde(rename = "languageAlias")]
+    pub language_aliases: LiteMap<String, Replacement<String>>,
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/serde/aliases.rs:23]: Praise (here and elsewhere): thanks for fixing this

File: provider/datagen/src/transform/cldr/serde/numbers.rs [hunk 6]
```diff
@@ -96,7 +95,7 @@ pub struct Numbers {
     #[serde(rename = "defaultNumberingSystem")]
     pub default_numbering_system: TinyStr8,
     #[serde(rename = "minimumGroupingDigits")]
-    #[serde(deserialize_with = "deserialize_number_from_string")]
+    #[serde(deserialize_with = "serde_aux::prelude::deserialize_number_from_string")]
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/serde/numbers.rs:98]: Praise: good idea

File: provider/datagen/src/transform/uprops/enum_uniset.rs [hunk 7]
```diff
@@ -12,6 +14,7 @@ use icu_provider::prelude::*;
 use icu_uniset::UnicodeSetBuilder;
 use std::sync::RwLock;
 
+/// A data provider reading from TOML files produced by the ICU4C icuwriteuprops tool.
```
Comment (sffc) [inline provider/datagen/src/transform/uprops/enum_uniset.rs:17]: Nit:
```suggestion
/// A data provider reading from TOML files produced by the ICU4C icuexportdata tool.
```

File: provider/datagen/src/transform/uprops/mod.rs [hunk 8]
```diff
@@ -2,9 +2,11 @@
 // called LICENSE at the top level of the ICU4X source tree
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
-//! `icu_datagen::uprops` contains implementations of the [`ICU4X`]
+//! This module contains implementations of the [`ICU4X`]
 //! [data provider] interface backed by TOML files exported by the
-//! ICU4C icuwriteuprops tool. Create a directory containing TOML files for
+//! ICU4C icuwriteuprops tool.
```
Comment (sffc) [inline provider/datagen/src/transform/uprops/mod.rs:7]: Nit (here and elsewhere): I know this is pre-existing but the tool is named "icuexportdata" not "icuwriteuprops" (this was an older name for the tool when it was being drafted)

File: provider/datagen/src/transform/uprops/script.rs [hunk 9]
```diff
@@ -16,11 +16,11 @@ use icu_provider::prelude::*;
 use std::convert::TryFrom;
 use zerovec::{VarZeroVec, ZeroSlice, ZeroVec};
 
+/// A data provider reading from TOML files produced by the ICU4C icuwriteuprops tool.
```
Comment (sffc) [inline provider/datagen/src/transform/uprops/script.rs:19]: Nit: `icuexportdata`

**Review by sffc — CHANGES_REQUESTED** [review]
LGTM overall, just some small things

**Review by sffc — APPROVED** [review]

**Pull request merged**